smartsim/_core/mli/workermanager.py

```diff
@@ -0,0 +1,976 @@
+        """
+
+    @abstractmethod
+    def _can_shutdown(self) -> bool:
+        """Return true when the criteria to shut down the service are met."""
+
+    def _on_start(self) -> None:
+        """Empty hook method for use by subclasses. Called on initial entry into
+        ServiceHost `execute` event loop before `_on_iteration` is invoked."""
+        logger.debug(f"Starting {self.__class__.__name__}")
+
+    def _on_shutdown(self) -> None:
+        """Empty hook method for use by subclasses. Called immediately after exiting
+        the main event loop during automatic shutdown."""
+        logger.debug(f"Shutting down {self.__class__.__name__}")
+
+    def _on_cooldown(self) -> None:
+        """Empty hook method for use by subclasses. Called on every event loop
+        iteration immediately upon exceeding the cooldown period"""
+        logger.debug(f"Cooldown exceeded by {self.__class__.__name__}")
+
+    def execute(self) -> None:
+        """The main event loop of a service host. Evaluates shutdown criteria and
+        combines with a cooldown period to allow automatic service termination.
+        Responsible for executing calls to subclass implementation of `_on_iteration`"""
+        self._on_start()
+
+        start_ts = time.time_ns()
+        last_ts = start_ts
+        running = True
+        elapsed_cooldown = 0
+        nanosecond_scale_factor = 1000000000
+        cooldown_ns = self._cooldown * nanosecond_scale_factor
+
+        # if we're run-once, use cooldown to short circuit
+        if not self._as_service:
+            self._cooldown = 1
+            last_ts = start_ts - (cooldown_ns * 2)
+
+        while running:
+            self._on_iteration(start_ts)
+
+            eligible_to_quit = self._can_shutdown()
+
+            if self._cooldown and not eligible_to_quit:
+                # reset timer any time cooldown is interrupted
+                elapsed_cooldown = 0
+
+            # allow service to shutdown if no cooldown period applies...
+            running = not eligible_to_quit
+
+            # ... but verify we don't have remaining cooldown time
+            if self._cooldown:
+                elapsed_cooldown += start_ts - last_ts
+                remaining = cooldown_ns - elapsed_cooldown
+                running = remaining > 0
+
+                rem_in_s = remaining / nanosecond_scale_factor
+
+                if not running:
+                    cd_in_s = cooldown_ns / nanosecond_scale_factor
+                    logger.info(f"cooldown {cd_in_s}s exceeded by {abs(rem_in_s):.2f}s")
+                    self._on_cooldown()
+                    continue
+
+                logger.debug(f"cooldown remaining {abs(rem_in_s):.2f}s")
+
+            last_ts = start_ts
+            start_ts = time.time_ns()
+
+            if self._loop_delay:
+                time.sleep(abs(self._loop_delay))
+
+        self._on_shutdown()
+
+
+class WorkerManager(ServiceHost):
```

Comment:
Could just be personal preference, but ``Service`` might be a better class name.  But maybe I'm also missing what the ``Host`` clarifies.

Reply:
Host was meant to imply "I am an application that hosts some implementation of a service" but I am not strongly attached to it. Renamed as `Service`